fbrp/src/fbrp/runtime/conda.py

```diff
@@ -76,19 +77,46 @@ def __init__(
         self.args = args
 
     async def run(self):
-        self.set_state(BaseLauncher.State.STARTING)
-        self.proc = await asyncio.create_subprocess_shell(
+        life_cycle.set_state(self.name, life_cycle.State.STARTING)
+
+        # We grab the conda env variables separate from executing the run
+        # command to simplify detecting pid and removing some race conditions.
+
+        envinfo = await asyncio.create_subprocess_shell(
             f"""
                 eval "$(conda shell.bash hook)"
                 conda activate fbrp_{self.name}
-                cd {self.proc_def.root}
-                {shlex.join(self.run_command)}
+                echo "fbrp env begin"
+                env
+                echo "fbrp env end"
             """,
             stdout=asyncio.subprocess.PIPE,
+            executable="/bin/bash",
+            cwd=self.proc_def.root,
+        )
+
+        # read blocks until EOF.
+        lines = (await envinfo.stdout.read()).decode().split("\n")
+        lines = lines[lines.index("fbrp env begin") + 1 : lines.index("fbrp env end")]
+        # Multiline functions may be defined with the form "FOO%%=() { ...}"
+        # For now, try to ignore those.
+        # TODO(lshamis): Find a better way to parse/extract env.
+        lines = [l for l in lines if "=" in l and "%%=" not in l]
+        conda_env = dict(line.split("=", 1) for line in lines)
```

Comment:
Would it be simpler to do something like this ?

- Export all env variables to a file

```
        envinfo = await asyncio.create_subprocess_shell(
            f"""
                eval "$(conda shell.bash hook)"
                conda activate fbrp_{self.name}
                echo "fbrp env begin"
                set > /tmp/fbrp_{self.name}_env.txt
                echo "fbrp env end"
            """,
            stdout=asyncio.subprocess.PIPE,
            executable="/bin/bash",
            cwd=self.proc_def.root,
        )
```

Then parse the file:

```
lines = await envinfo.wait()
conda_env = <parse contents of /tmp/fbrp_{self.name}_env.txt>
```